### PR TITLE
Add support for dispatching blockmap to segment servers.

### PR DESCRIPTION
**Pull request by higuoxing**

### CODE REVIEW

**Review by xuebinsu — APPROVED** [review]
+1

File: quotamodel.c [hunk 1]
```diff
@@ -287,6 +305,7 @@ add_quota_to_blacklist(QuotaType type, Oid targetOid, Oid tablespaceoid, bool se
 	keyitem.databaseoid = MyDatabaseId;
 	keyitem.tablespaceoid = tablespaceoid;
 	keyitem.targettype = (uint32) type;
+	memset(&keyitem.relfilenode, 0, sizeof(RelFileNode));
```
Comment (zhrt123) [inline quotamodel.c:308]: this line can be remove

File: quotamodel.c [hunk 2]
```diff
@@ -1334,40 +1353,56 @@ get_rel_owner_schema_tablespace(Oid relid, Oid *ownerOid, Oid *nsOid, Oid *table
...
+	memcpy(&keyitem.relfilenode, &relfilenode, sizeof(RelFileNode));
+
+	LWLockAcquire(diskquota_locks.black_map_lock, LW_SHARED);
+	entry = hash_search(disk_quota_black_map,
+						&keyitem, HASH_FIND, &found);
+
+	if (found && entry)
+	{
+		GlobalBlackMapEntry segblackentry;
+		memcpy(&segblackentry.keyitem, &entry->auxblockinfo, sizeof(BlackMapEntry));
```
Comment (zhrt123) [inline quotamodel.c:1375]: copy entry->segexceeded to segblackentry

File: quotamodel.c [hunk 3]
```diff
@@ -1476,3 +1541,456 @@ export_exceeded_error(GlobalBlackMapEntry *entry)
...
+		lt = DatumGetHeapTupleHeader(datums[i]);
+		keyitem.targetoid     = DatumGetObjectId(GetAttributeByNum(lt, 1, &isnull));
+		keyitem.databaseoid   = DatumGetObjectId(GetAttributeByNum(lt, 2, &isnull));
+		keyitem.tablespaceoid = DatumGetObjectId(GetAttributeByNum(lt, 3, &isnull));
+		keyitem.targettype    = DatumGetInt32(GetAttributeByNum(lt, 4, &isnull));
+		segexceeded           = DatumGetBool(GetAttributeByNum(lt, 5, &isnull));
+
+		blackmapentry = hash_search(disk_quota_black_map, &keyitem, HASH_ENTER_NULL, NULL);
+		if (blackmapentry)
+			blackmapentry->segexceeded = segexceeded;
```
Comment (zhrt123) [inline quotamodel.c:1713]: add blackmapentry->keyitem = keyitem;

File: quotamodel.c [hunk 4]
```diff
@@ -1476,3 +1541,456 @@ export_exceeded_error(GlobalBlackMapEntry *entry)
...
+							BlackMapEntry				blocked_filenode_keyitem;
+
+							memset(&blocked_filenode_keyitem, 0, sizeof(BlackMapEntry));
+							memcpy(&blocked_filenode_keyitem.relfilenode, &relfilenode, sizeof(RelFileNode));
+
+							blocked_filenode_entry = hash_search(disk_quota_black_map,
+																 &blocked_filenode_keyitem,
+																 HASH_ENTER_NULL, &found);
+
+							if (!found && blocked_filenode_entry)
```
Comment (zhrt123) [inline quotamodel.c:1835]: update blocked_filenode_entry->segexceeded

**Comment by higuoxing** [comment]
I'm running out of time today. Will add some tests tomorrow : p

**Review by xuebinsu — APPROVED** [review]
+1

**Pull request merged**